rust/src/storage/s3/mod.rs

```diff
@@ -405,80 +453,91 @@ pub trait LockClient: Send + Sync + Debug {
     /// successfully released, and false if someone else already stole the lock
     async fn release_lock(&self, lock: &LockItem) -> Result<bool, StorageError>;
 }
 
-#[allow(clippy::unnecessary_wraps)]
-fn try_create_lock_client(_region: Region) -> Result<Option<Box<dyn LockClient>>, StorageError> {
-    match std::env::var("AWS_S3_LOCKING_PROVIDER") {
-        Ok(p) if p.to_lowercase() == "dynamodb" => {
-            cfg_if::cfg_if! {
-                if #[cfg(feature = "dynamodb")] {
-                    let client = dynamodb_lock::DynamoDbLockClient::new(
-                        rusoto_dynamodb::DynamoDbClient::new(_region),
-                        dynamodb_lock::Options::default()
-                    );
-                    Ok(Some(Box::new(client)))
-                } else {
-                    Err(StorageError::Generic("dynamodb feature is not enabled".to_string()))
-                }
+const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+
+impl dyn LockClient {
+    async fn rename_with_lock(
+        &self,
+        s3: &S3StorageBackend,
+        src: &str,
+        dst: &str,
+    ) -> Result<(), StorageError> {
+        let mut lock = self.acquire_lock_loop(src, dst).await?;
+
+        if let Some(ref data) = lock.data {
+            let data: LockData = serde_json::from_str(data)
+                .map_err(|_| StorageError::S3Generic("Lock data deserialize error".to_string()))?;
+
+            let mut rename_result = s3.unsafe_rename_obj(&data.source, &data.destination).await;
+
+            if rename_result.is_ok() && lock.acquired_expired_lock {
+                // If we acquired expired lock then the rename done above is
+                // a repair of expired one. So on this time we try the intended rename.
+
+                // lock.data = LockData { src, dst }
+                // s3.update_data()
+                lock.data = Some(LockData::json(src, dst)?);
+                lock = self.update_data(&lock).await?;
+                rename_result = s3.unsafe_rename_obj(src, dst).await;
             }
-        }
-        _ => Ok(None),
-    }
-}
 
-const DEFAULT_MAX_RETRY_ACQUIRE_LOCK_ATTEMPTS: u32 = 10_000;
+            let release_result = self.release_lock(&lock).await;
+
+            // before unwrapping `rename_result` the `release_result` is called to ensure that we
+            // no longer hold the lock
+            rename_result?;
 
-async fn rename_with_lock(
-    s3_backend: &S3StorageBackend,
-    src: &str,
-    dst: &str,
-) -> Result<(), StorageError> {
-    let lock_client = match s3_backend.lock_client {
-        Some(ref lock_client) => lock_client,
-        None => {
-            return Err(StorageError::S3Generic(
-                "dynamodb locking is not enabled".to_string(),
+            if !release_result? {
+                log::error!("Could not release lock {:?}", &lock);
+                return Err(StorageError::S3Generic("Lock is not released".to_string()));
```

Comment:
cool, in this case, i think it might be better to return Ok with a state to signal different success cases. That way for callers who don't care about the details, they can just match on `Ok(_)`. Otherwise, I am a little bit concerned that some callers could introduce bugs by always assuming getting `Err(_)` means a rename failure.

Reply:
e.g. like a bool value? I saw that is a common practice for rename operations in some APIs

Reply:
Still, when thinking on this with perspective from https://github.com/delta-io/delta-rs/pull/233#discussion_r632431144 I feel like we might leave it as is here too, like yes, the rename succeeded but we still got unexpected error, which might involve to take some action, simply silent it and leave as is might lead into unexpected behaviour

Reply:
In an API, typically this is done by returning Ok with a enum state to signal different cases. For example: `Ok(RenameResult::LockPreempted)`, `Ok(RenameResult::NoError)`, etc.

Then for applications that are OK with pauses, they can just match on `Ok(_)`. For applications that cannot live with pauses, they can explicitly match on `Ok(RenameResult::NoError)`:

```rust
match rename_with_lock() {
    Ok(RenameResult::NoError) => {},
    Ok(_) => {
        // Handle pauses 
    }
    Err(_) => // handle rename did not go through,
}
```

`Err` usually signals unrecoverable errors, so it's more likely that a client may assume this and check `Err(_)` for error, then proceed with retrying the rename with the same data even after the rename actually went through if we wrap this `lock is not released` error in `Err`.

Reply:
So at this point we assume that delta-rs write can live with pauses? E.g. wee still need to handle `RenameResult` into err or ok within try_commit loop

Reply:
Btw, feels like we might expose an API as you described and leave it up to user on how to deal with it, through some config flag

Reply:
Yeah, I think delta-rs should be able to support use-cases that are ok with pauses and let the user decide whether they want to crash immediately on pause or continue as normal. Handling pauses automatically is a common design decision in distributed system designs.